### PR TITLE
Adds machine breakdowns and repairs (UPDATED!)

### PR DESCRIPTION
**Pull request by Goldsteel**
[discussion][wip][gameplay][tweak]

Sounds and icons are currently temporary and will be updated with something better than the cyborg component and windows crash sound. Don't worry.

Current features:
(1) Machines now have HP! This is currently 100/100.
(2) Machines can now be damaged! Shooting, EMPing and otherwise abusing your kit will degrade it's HP.
(3) When the machine HP hits 0, it undergoes the current component breakdown process, coughs and splutters and grinds to a halt.
(4) Are you a mechanic? No? Good. You'll still need nanopaste to heal your equipment, though. Nanopaste will restore machine HP without you needing to go through breakdowns.
(5) Examine machines to see if they're quite damaged.

:cl:
 * rscadd: Machines now have HP and will breakdown at 0! This will deplete if you damage them. They will spark, splutter and grind to a halt.
 * rscadd: If a machine isn't entirely broken, use some nanopaste to restore it's HP.
 * rscadd:  Broken machines contain a broken part. Replace it to fix it!
 * rscadd: The RPED and it's upgraded version can be used to quickly repair broken parts.

### CODE REVIEW

File: code/modules/power/power.dm [hunk 1]
```diff
@@ -144,6 +144,8 @@
 // by default, check equipment channel & set flag
 // can override if needed
 /obj/machinery/proc/power_change()
+	if(BROKEN) // See machine breakdowns in machinery/machinery.dm
```
Comment (MadmanMartian) [inline code/modules/power/power.dm]: ???

**Comment by Kharkov72** [comment]
Reposting that I would rather it be a new random event / something tha happens when it's fucked with (EMPs?) rather than something that can just randomly happen, they're mass produced machines in a time where space travel is normal, they wouldn't just randomly break.

File: code/modules/power/power.dm (hunk 1 above)
Comment (Goldsteel) [reply code/modules/power/power.dm]: oversight, removed

**Comment by aceedex** [comment]
Neat idea that isn't very infringent on mechanics' more laid back approach.
All I really have to note is that SMES should have a damn high break tolerance, and that machines shouldn't become completely useless if they break to save some headache for those 5 man 4am rounds and should instead severely lose efficiency

**Comment by Goldsteel** [comment]
The idea of machines becoming less efficient is on the table. This is just one implementation of the idea. You can set custom break chances (or 0 for never) for machines like the SMES and telecomms.

There is an ongoing issue with meme machines like cryo and vending machines, but I'm tired so that's for tomorrow.

**Comment by DrSnips** [comment]
No

**Comment by BomberBro** [comment]
I'd be all for maybe weapons being able to damage machinery (shooting up an autolathe or toolboxing a console to break it), but having them just break for no reason or even a gay random event is just pointless busywork for the sake of busywork.

**Comment by ghost** [comment]
Add a flag to disable this for certain machines and have it be part of ion storm and electrical overload instead of a per-machine basis.

**Comment by SonixApache** [comment]
testmerge

**Comment by FudgePucker** [comment]
Awful fucking idea.

First off the only jobs with access to spare parts are notorious shutters. The mechanic was lynched for spare rushing, the scientists will ignore you, and cargo has declared independence. Guess you gotta break into secure tech storage.

Secondly this will make the wild ride even more unbearable. Oh boy the cyrotube with the cyromix broke, time to be shoved into the empty cyrotube until the round ends.

Finally, mechanics are fucking awful and relying on a job filled with shitters just because of random chance bullshit is fucking stupid. It's like expecting assistants to fucking assist.

**Comment by MadmanMartian** [comment]
Fudge, the cryo cell can't break as it's on the atmos machine sub, not the main machine sub, so in this itteration it can't break down.

**Comment by Goldsteel** [comment]
I've taken on Bomberbros advice. I am now in the process of coding shooting/kicking/attacking machines having a chance of breakdown. Machines no longer break randomly, they must be smashed.

EDIT: I'll also probably add in something about EMPs? Not sure yet.

**Comment by FudgePucker** [comment]
Kicking machines breaking componants will anger the science meta club because the fastest way to deconstruct stuff is to kick it.

**Comment by Goldsteel** [comment]
I can always just set the DA to be breakdown 0 and put a meme in it's description about "Having a kickplate"

**Comment by Goldsteel** [comment]
I'll be putting this on hold while r0b0 implements a system for machine health. Hopefully it will make more sense to have machines hold a HP value that decreases when they're damaged/used (perhaps healing machines with nanopaste?). If a machine drops to 0 HP then the breakdowns will fire.

**Comment by Goldsteel** [comment]
Machines now have a HP value that decreases when they're shot or kicked (kick damage is only 1, don't lose your shit just yet). If a machine is just worn, you can apply nanopaste to it to give it it's HP back to 100.

Here's some examples of this working:

Nanopaste repairs to non-broken down machines.
image
image

Doing damage to machines makes them break down with no RNG involved, currently.
image

Offending components can be removed with an RPED/BRPED and replaced, or done manually a lot slower. This restores the machine to full HP.
image
image

I'm working on dealing with all the snowflake machines and vendors, but it shouldn't be too difficult. 

COMING SOON: Snowflake machine fixes, visual indication of breakdown activity.

**Comment by Goldsteel** [comment]
So, it turns out computers are a machine subtype and this breaks them. I'll have to fix that too. Vending machines also won't update their icons if they're destroyed and repaired. I didn't want there to be snowflake code for every machine, but almost every single machine has some degree of snowflake code.

**Comment by tuypo1** [comment]
Yeah making it a random announced event instead of just a random chance on the machine itself seems like it would be a better implementation

**Comment by Goldsteel** [comment]
It's no longer random. It's now a machine HP based system. Damage machines by kicking, shooting or meleeing them if the object does enough damage.

I managed to get vending machines and consoles to stop being shit.

**Comment by Goldsteel** [comment]
So, it turns out, lots of consoles have unique attack_hand() overrides and many common things also are machines with no parts like airlocks, shutters, air alarms, light switchs and so on. This will take more time to reconcile.

File: code/game/machinery/machinery.dm [hunk 2]
```diff
@@ -427,6 +452,10 @@ Class Procs:
 	if(stat & (NOPOWER|BROKEN|MAINT))
 		return 1
 
+	if(stat & (BROKEN))
+		to_chat(user, "<span class='warning'>\The [src] is broken.</span>")
+		return
```
Comment (PaulRitter) [inline code/game/machinery/machinery.dm]: this should be return 1 no?

Comment (PaulRitter) [inline code/game/machinery/machinery.dm:457]: this should be `return 1` no?

File: code/game/machinery/machinery.dm [hunk 3]
```diff
@@ -453,6 +482,19 @@ Class Procs:
 	src.add_fingerprint(user)
 	return 0
 
+///obj/machinery/attack_hand(mob/user as mob)
```
Comment (PaulRitter) [inline code/game/machinery/machinery.dm:485]: ?

File: code/game/machinery/machinery.dm [hunk 4]
```diff
@@ -201,6 +212,20 @@ Class Procs:
...
+obj/machinery/bullet_act(var/obj/item/projectile/Proj)
+	if(istype(Proj ,/obj/item/projectile/beam)||istype(Proj,/obj/item/projectile/bullet)||istype(Proj,/obj/item/projectile/ricochet))
+		if(!istype(Proj ,/obj/item/projectile/beam/lasertag) && !istype(Proj ,/obj/item/projectile/beam/practice) )
+			log_attack("<font color='red'>[key_name(Proj.firer)] shot [src]/([formatJumpTo(src)]) with a [Proj.type]</font>")
+			if(Proj.firer)//turrets don't have "firers"
+				Proj.firer.attack_log += "\[[time_stamp()]\] <b>[key_name(Proj.firer)]</b> shot <b>[src]([x],[y],[z])</b> with a <b>[Proj.type]</b>"
+			if(machine_health == null)
+				return
+			if(machine_health >= 0)
+				machine_health = max(0,machine_health - Proj.damage)
```
Comment (PaulRitter) [inline code/game/machinery/machinery.dm:224]: would be cleaner to make a obj/machinery/take_damage and have these two lines in there. Would also be more future proof (ex. Machines saying ouch when being damaged. i dunno)

File: code/game/machinery/machinery.dm [hunk 5]
```diff
@@ -758,5 +810,34 @@ Class Procs:
 			visible_message("<span class='notice'>\A [scan] pops out of \the [src]!</span>")
 			scan = null
 
+	if(!stat & (BROKEN)) // If it's not broken, apply some damage to it.
+		if(machine_health == null)
+			return
+		if(machine_health >=0)
+			machine_health = max(0,machine_health - 1)
```
Comment (PaulRitter) [inline code/game/machinery/machinery.dm:817]: would be cleaner to make a obj/machinery/take_damage and have these two lines in there. Would also be more future proof (ex. Machines saying ouch when being damaged. i dunno)

File: code/game/machinery/computer/security.dm [hunk 6]
```diff
@@ -39,6 +39,9 @@
 
 //Someone needs to break down the dat += into chunks instead of long ass lines.
 /obj/machinery/computer/secure_data/attack_hand(mob/user as mob)
+	if(stat & (BROKEN))
```
Comment (PaulRitter) [inline code/game/machinery/computer/security.dm:42]: wont be needed if you do the `return 1` in `obj/machinery/attack_hand`

File: code/game/machinery/computer/arcade/arcade.dm [hunk 7]
```diff
@@ -41,6 +41,9 @@
 	A.game_data = game.export_data()
 
 /obj/machinery/computer/arcade/attack_hand(mob/user)
+	if(stat & (BROKEN))
```
Comment (PaulRitter) [inline code/game/machinery/computer/arcade/arcade.dm:44]: wont be needed if you do the `return 1` in `obj/machinery/attack_hand`

**Review by PaulRitter — CHANGES_REQUESTED** [review]

**Comment by Goldsteel** [comment]
This is still a WiP. There's a lot of stuff that's just hanging around that needs to be cleaned up.

**Comment by Kurfursten** [comment]
wip...

**Comment by Kurfursten** [comment]
Ask any collab to reopen this if you do work on it

**Pull request closed without merging**